// botframework-webchat-component/lib/Middleware/createTypingIndicatorRenderer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createTypingIndicatorRenderer;

var _react = _interopRequireDefault(require("react"));

var _concatMiddleware = _interopRequireDefault(require("./concatMiddleware"));

var _createCoreMiddleware = _interopRequireDefault(require("./TypingIndicator/createCoreMiddleware"));

var _ErrorBox = _interopRequireDefault(require("../ErrorBox"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint no-magic-numbers: ["error", { "ignore": [2] }] */
function createTypingIndicatorRenderer(additionalMiddleware) {
  var typingIndicatorMiddleware = (0, _concatMiddleware.default)(additionalMiddleware, (0, _createCoreMiddleware.default)())({});
  return function () {
    try {
      return typingIndicatorMiddleware(function (_ref) {
        var activeTyping = _ref.activeTyping,
            typing = _ref.typing,
            visible = _ref.visible;
        return /*#__PURE__*/_react.default.createElement(_ErrorBox.default, {
          message: "No renderer for typing indicator"
        }, /*#__PURE__*/_react.default.createElement("pre", null, JSON.stringify({
          activeTyping: activeTyping,
          typing: typing,
          visible: visible
        }, null, 2)));
      }).apply(void 0, arguments);
    } catch (err) {
      var message = err.message,
          stack = err.stack;
      console.error({
        message: message,
        stack: stack
      });
      return /*#__PURE__*/_react.default.createElement(_ErrorBox.default, {
        error: err,
        message: "Failed to render typing indicator"
      }, /*#__PURE__*/_react.default.createElement("pre", null, JSON.stringify({
        message: message,
        stack: stack
      }, null, 2)));
    }
  };
}
