// botframework-webchat-component/lib/ScreenReaderText.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _glamor = require("glamor");

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint react/forbid-dom-props: ["off"] */
var ROOT_CSS = (0, _glamor.css)({
  // .sr-only - This component is intended to be invisible to the visual Web Chat user, but read by the AT when using a screen reader
  color: 'transparent',
  height: 1,
  overflow: 'hidden',
  position: 'absolute',
  // We need to set top: 0, otherwise, it will repro:
  // - Run NVDA
  // - Make the transcript long enough to show the scrollbar
  // - Press SHIFT-TAB, focus on upload button
  // - Press up arrow multiple times
  top: 0,
  whiteSpace: 'nowrap',
  width: 1
});

var ScreenReaderText = function ScreenReaderText(_ref) {
  var id = _ref.id,
      text = _ref.text;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)(ROOT_CSS + ''),
    id: id
  }, text);
};

ScreenReaderText.defaultProps = {
  id: undefined
};
ScreenReaderText.propTypes = {
  id: _propTypes.default.string,
  text: _propTypes.default.string.isRequired
};
var _default = ScreenReaderText;
exports.default = _default;
