// botframework-webchat-component/lib/Middleware/Attachment/createCoreMiddleware.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createCoreMiddleware;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _AudioAttachment = _interopRequireDefault(require("../../Attachment/AudioAttachment"));

var _FileAttachment = _interopRequireDefault(require("../../Attachment/FileAttachment"));

var _ImageAttachment = _interopRequireDefault(require("../../Attachment/ImageAttachment"));

var _TextAttachment = _interopRequireDefault(require("../../Attachment/TextAttachment"));

var _VideoAttachment = _interopRequireDefault(require("../../Attachment/VideoAttachment"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// TODO: [P4] Rename this file or the whole middleware, it looks either too simple or too comprehensive now
function createCoreMiddleware() {
  return function () {
    return function (next) {
      var Attachment = function Attachment(_ref) {
        var _ref$activity = _ref.activity,
            activity = _ref$activity === void 0 ? {} : _ref$activity,
            _ref$activity2 = _ref.activity;
        _ref$activity2 = _ref$activity2 === void 0 ? {} : _ref$activity2;
        var _ref$activity2$from = _ref$activity2.from;
        _ref$activity2$from = _ref$activity2$from === void 0 ? {} : _ref$activity2$from;
        var role = _ref$activity2$from.role,
            attachment = _ref.attachment,
            _ref$attachment = _ref.attachment;
        _ref$attachment = _ref$attachment === void 0 ? {} : _ref$attachment;
        var contentType = _ref$attachment.contentType,
            contentUrl = _ref$attachment.contentUrl,
            thumbnailUrl = _ref$attachment.thumbnailUrl;
        return role === 'user' && !/^text\//.test(contentType) && !thumbnailUrl ? /*#__PURE__*/_react.default.createElement(_FileAttachment.default, {
          activity: activity,
          attachment: attachment
        }) : /^audio\//.test(contentType) ? /*#__PURE__*/_react.default.createElement(_AudioAttachment.default, {
          activity: activity,
          attachment: attachment
        }) : /^image\//.test(contentType) ? /*#__PURE__*/_react.default.createElement(_ImageAttachment.default, {
          activity: activity,
          attachment: attachment
        }) : /^video\//.test(contentType) ? /*#__PURE__*/_react.default.createElement(_VideoAttachment.default, {
          activity: activity,
          attachment: attachment
        }) : contentUrl || contentType === 'application/octet-stream' ? /*#__PURE__*/_react.default.createElement(_FileAttachment.default, {
          activity: activity,
          attachment: attachment
        }) : /^text\//.test(contentType) ? /*#__PURE__*/_react.default.createElement(_TextAttachment.default, {
          activity: activity,
          attachment: attachment
        }) : next({
          activity: activity,
          attachment: attachment
        });
      };

      Attachment.propTypes = {
        activity: _propTypes.default.any.isRequired,
        attachment: _propTypes.default.shape({
          contentType: _propTypes.default.string.isRequired,
          contentUrl: _propTypes.default.string,
          thumbnailUrl: _propTypes.default.string
        }).isRequired
      };
      return Attachment;
    };
  };
}
