// react-say/lib/createSynthesize.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = createSynthesize;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _QueuedUtterance = _interopRequireDefault(require("./QueuedUtterance"));

function createSynthesize() {
  var queueWithCurrent = [];
  var running;

  var run = function run() {
    return _regenerator["default"].async(function run$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            if (!running) {
              _context2.next = 2;
              break;
            }

            return _context2.abrupt("return");

          case 2:
            running = true;
            _context2.prev = 3;
            _context2.next = 6;
            return _regenerator["default"].awrap(function _callee() {
              var queuedUtterance;
              return _regenerator["default"].async(function _callee$(_context) {
                while (1) {
                  switch (_context.prev = _context.next) {
                    case 0:
                      if (!(queuedUtterance = queueWithCurrent[0])) {
                        _context.next = 12;
                        break;
                      }

                      _context.prev = 1;
                      _context.next = 4;
                      return _regenerator["default"].awrap(queuedUtterance.speak());

                    case 4:
                      _context.next = 9;
                      break;

                    case 6:
                      _context.prev = 6;
                      _context.t0 = _context["catch"](1);
                      // TODO: If the error is due to Safari restriction on user touch
                      //       The next loop on the next audio will also fail because it was not queued with a user touch
                      _context.t0.message !== 'cancelled' && console.error(_context.t0);

                    case 9:
                      queueWithCurrent = queueWithCurrent.filter(function (target) {
                        return target !== queuedUtterance;
                      });
                      _context.next = 0;
                      break;

                    case 12:
                    case "end":
                      return _context.stop();
                  }
                }
              }, null, null, [[1, 6]]);
            }());

          case 6:
            _context2.prev = 6;
            running = false;
            return _context2.finish(6);

          case 9:
          case "end":
            return _context2.stop();
        }
      }
    }, null, null, [[3,, 6, 9]]);
  };

  return function (ponyfill, utterance) {
    var _ref = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {},
        onEnd = _ref.onEnd,
        onError = _ref.onError,
        onStart = _ref.onStart;

    if (!(utterance instanceof ponyfill.SpeechSynthesisUtterance)) {
      throw new Error('utterance must be instance of the ponyfill');
    }

    var queuedUtterance = new _QueuedUtterance["default"](ponyfill, utterance, {
      onEnd: onEnd,
      onError: onError,
      onStart: onStart
    });
    queueWithCurrent = [].concat((0, _toConsumableArray2["default"])(queueWithCurrent), [queuedUtterance]);
    run();
    return {
      // The cancel() function returns a Promise
      cancel: function cancel() {
        return queuedUtterance.cancel();
      },
      promise: queuedUtterance.promise
    };
  };
}
